### PR TITLE
Add support for sampling to stdio logger

### PR DESCRIPTION
**Pull request by douglas-reid**
This PR adds sampling to the stdio logger. This has been a highly-requested feature
for some time now, especially given performance concerns with excessive I/O and the 
overall volume of logs generated in high-request rate environments.

This PR uses the `zapcore.sampler` out-of-the-box to achieve rate-limited
logs. This has a number of interesting properties:
- using `request.time` as the time for comparisons can impact sampling of out-of-order Reports
- stdio currently doesn't set any `Message` for logs, meaning all logs for a given handler
  will be counted by a single bucket
- any tailoring based on attributes / variables *must* be done at the handler config level.

Possible Future Enhancements:
We could reimplement the sampler code to add `Field` awareness and/or enable passing of Report()
time in addition to the `instance.Timestamp`.

### CODE REVIEW

File: mixer/adapter/stdio/config/config.proto [hunk 1]
```diff
@@ -118,4 +119,23 @@ message Params {
     // The maximum number of old rotated log files to retain.  The default
     // is to retain at most 1000 logs. 0 indicates no limit.
     int32 max_rotated_files = 9;
+
+    // Controls the level of sampling that will be used by a stdio handler for logs entries.
```
Comment (geeknoid) [inline mixer/adapter/stdio/config/config.proto]: logs entries -> log entries

File: mixer/adapter/stdio/config/config.proto [hunk 2]
```diff
@@ -118,4 +119,23 @@ message Params {
...
+    // Sampling is configured based on a duration, with the following logic:
+    //  - over the course of the duration, `max_unsampled_entries` are allowed unmodified
+    //  - after `max_unsampled_entries`, for the remainder of the duration a single log entry will
+    //    be emitted for every `sampling_rate` entries received.
+    message LogsSampling {
```
Comment (geeknoid) [inline mixer/adapter/stdio/config/config.proto]: I'm not liking the S in there. I think LogSampling would be better.

Comment (douglas-reid) [reply mixer/adapter/stdio/config/config.proto]: renamed everything to remove the 's'(es).

**Comment by douglas-reid** [comment]
What do we think of this approach?  Reasonable?  Do we care about making it harder to misuse (`Report()`s with wildly varying `request.time`s being handled in batch, etc.) ?

**Comment by geeknoid** [comment]
This seems reasonable as a first cut if we want sampling.

However, I wonder if the issue is really that we're not doing any buffering on the log stream. So it ends up being very expensive to log every single entry,. If we just put in a few K of buffer space in there, we could dramatically reduce the overall cost and then we could probably log everything that comes in with minimal cost.

**Comment by douglas-reid** [comment]
I think buffering could be added on top. IIRC, there were complaints about both the I/O time, as well as the general volume of logs.

**Comment by costinm** [comment]
Ping ? 
@mandarjog  can we check how much this helps in "mixer not crashing", and maybe backport to 1.0  ?

**Comment by geeknoid** [comment]
I think we should put generic support for sampling of telemetry in Mixer, rather than in individual adapters. WDYT?

**Comment by douglas-reid** [comment]
closing this in favor of an implementation based on the recent api changes.

**Pull request closed without merging**